### PR TITLE
Style up Little Machine results.

### PR DESCRIPTION
**Pull request by chuckharmston**
screen shot 2015-11-23 at 11 53 51 am

r? @6a68 
#### Known issues
1. Tiny Machine emboldens exact query matches. [I don't think it should do that](https://github.com/nchapman/tiny_machine/issues/2), or at least not by sending HTML over the wire.
2. It's a little wasteful to bundling icons for both states. There are other approaches that would both be slimmer, and let us transition those colors. https://github.com/mozilla/universal-search-content/issues/75
3. We probably [don't need to base64 SVGs at all](https://css-tricks.com/probably-dont-base64-svg/).
4. There are a bunch of drive-by/organizational changes in here. I'm still unfamiliar enough with the ecosystem that I'm not doing a very good job planning my work. That should get better.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -29,6 +29,8 @@
     "raw-loader": "^0.5.1",
     "sass-loader": "^1.0.2",
     "style-loader": "^0.12.3",
+    "svg-inline-loader": "^0.4.0",
+    "svg-url-loader": "^1.0.0",
```
Comment (chuckharmston) [inline package.json]: - [x] These need to be removed. Artifact of failed experiments.

File: src/models/activity_result.js [hunk 2]
```diff
@@ -1,9 +1,15 @@
 import State from 'ampersand-state';
 
+
 export default State.extend({
```
Comment (chuckharmston) [inline src/models/activity_result.js:3]: - [x] Snuck some extra whitespace in there.

File: src/styles/main.scss [hunk 3]
```diff
@@ -29,72 +43,172 @@ body {
...
+      &.result--segment--little-machine {
+        display: flex;
+      }
     }
 
-    &:hover {
-      background: $border-grey;
+    .result--segment + .result--segment:before {
+      color: $color--grey;
+      content: '\A0\2014';
```
Comment (chuckharmston) [inline src/styles/main.scss]: - [x] `\A0` is a nonbreaking space. We should have one on both sides.

File: src/templates/little_machine/item.html [hunk 4]
```diff
@@ -1,12 +1,11 @@
...
-      {{{model.title}}} &mdash;
-    {{/model.title}}
-    <span class="result-url">
-      {{model.url}}
-    </span>
+<li class="activity-item result result--little-machine">
+  <span class="icon icon--recommendation"></span>
+  <span class="result--segment result--segment--little-machine">
+    <figure class="little-machine--segment little-machine--image">
+      <img src="{{model.image.url}}" width="64" height="64"/>
```
Comment (pdehaan) [inline src/templates/little_machine/item.html]: Do we have feelings on hard coding `width` and `height` in HTML markup instead of CSS?

Comment (chuckharmston) [inline src/templates/little_machine/item.html]: - [x] We should definitely use CSS for this, and that should be a Sass variable.

**Comment by chuckharmston** [comment]
Added a fix for issue #79, which adds an inset shadow to the image:

screen shot 2015-11-23 at 12 55 24 pm
screen shot 2015-11-23 at 12 55 35 pm
screen shot 2015-11-23 at 12 56 00 pm

File: src/styles/main.scss [hunk 5]
```diff
@@ -29,72 +45,180 @@ body {
...
+
+.icon--recommendation {
+  background-image: url(../images/icons/recommendation.svg);
+
+  .result--selected & {
+    background-image: url(../images/icons/recommendation-selected.svg);
+  }
 }
 
 .icon {
```
Comment (pdehaan) [inline src/styles/main.scss:209]: Merge this `icon` selector with the one on line 195?

Comment (chuckharmston) [inline src/styles/main.scss:209]: That will happen once I've entirely moved what was existing (sprited PNGs) to what I'm changing it to (unsprited SVGs). I'm currently keeping them separated for clarity.

File: src/styles/main.scss [hunk 6]
```diff
@@ -29,72 +45,180 @@ body {
...
+        display: flex;
+      }
     }
 
-    &:hover {
-      background: $border-grey;
+    .result--segment + .result--segment:before {
+      color: $color--grey;
+      content: '\A0\2014\A0';
+      margin: 0 2px
```
Comment (pdehaan) [inline src/styles/main.scss]: nit: missing `;`

File: src/styles/main.scss (hunk 5 above)
Comment (pdehaan) [inline src/styles/main.scss:209]: Cool, I was just playing with [**sass-lint**](https://github.com/sasstools/sass-lint) and noticed a few warnings.

Not sure if it's worth adding to this repo. I wasn't able to get the vanilla version working, but it works fine if I use the **gulp-sass-lint** plugin, although adding Gulp just for that task may be a bit excessive. Currently reports about [53 warnings](https://gist.github.com/pdehaan/89175a6bbcf052c396ce), but some are silly and should be silenced.

**Comment by chuckharmston** [comment]
Bookmark icons now included in results:

screen shot 2015-11-23 at 1 10 12 pm

File: src/styles/main.scss (hunk 5 above)
Comment (chuckharmston) [inline src/styles/main.scss:209]: Maybe something like this? 

https://github.com/hyungjs/csslint-loader

File: src/styles/main.scss (hunk 6 above)
Comment (chuckharmston) [inline src/styles/main.scss]: - [x] Gone.

File: src/styles/main.scss (hunk 5 above)
Comment (pdehaan) [inline src/styles/main.scss:209]: Ah, for reasons that will never be clear to me, the `sass-lint` CLI returns zero output unless you add the `--verbose` flag. Seems weird, but OK.

So I added this to package.json:

``` js
"sass-lint": "sass-lint src/styles/**/*.scss --verbose",
```

Create the following `.sass-lint.yml` file:

``` yml
rules:
  force-attribute-nesting: 0
  force-element-nesting: 0
  force-pseudo-nesting: 0
  nesting-depth: [2, max-depth: 3]
  no-ids: 0
  no-qualifying-elements: 0
  quotes: 0
```

Run this code from my Terminal:

``` sh
npm run sass-lint
```

And then get this output (after I tweak a bit of Sass locally):

``` sh
✗ npm run sass-lint

> universal-search-content@0.0.1 sass-lint /Users/pdehaan/dev/tmp/del/universal-search-content
> sass-lint src/styles/**/*.scss --verbose


src/styles/main.scss
   36:3   warning  Expected `-moz-user-select`, found `box-sizing`                              property-sort-order
   37:3   warning  Expected `box-sizing`, found `-moz-user-select`                              property-sort-order
   37:3   warning  Vendor prefixes should not be used                                           no-vendor-prefixes
   73:17  warning  Indentation of 16, expected 4                                                indentation
   74:17  warning  Indentation of 16, expected 4                                                indentation
  158:27  warning  Color functions such as 'rgba' should only be used in variable declarations  no-color-literals
  220:1   warning  Rule `.icon` should be merged with the rule on line 197                      no-mergeable-selectors
  252:15  warning  Color literals such as 'eee' should only be used in variable declarations    no-color-literals
  253:28  warning  Color literals such as 'ccc' should only be used in variable declarations    no-color-literals
  257:12  warning  Color literals such as '555' should only be used in variable declarations    no-color-literals
  269:25  warning  Color literals such as 'ccc' should only be used in variable declarations    no-color-literals
  274:21  warning  Color literals such as 'ddd' should only be used in variable declarations    no-color-literals

✖ 12 problems (0 errors, 12 warnings)
```

Anyways, I'm _waaaayyy_ off topic here, so I'll file a separate bug or two and we can discuss elsewhere if this is useful.

---

**UPDATE:** Filed as #83

**Comment by jaredhirsch** [comment]
Nice work, Chuck! I have questions about some of the bigger refactorings / reorganizational changes; I'll ping you when I'm online tomorrow, and we can find time to talk those out.

I've spotted some styling issues, might feel better to make (some/all) of these followup bugs rather than fix in this PR, up to you:
- The image in the tiny machine results is obscured by the key or click focus:
    
- Broken images are, sadly, prominent in the UI when they happen. Maybe we should load the image with JS, and only show it if it loads ok? Something like the approach [discussed here](http://stackoverflow.com/questions/9714525/javascript-image-url-verify/9714891#9714891)? The problem is that we don't want an awkward empty white block next to some text, and we also don't want the janky feeling of shoving the text over to the right after the image loads. I'm not sure what the right answer is.
   
- The mouse focus and keyboard focus are intended to be independent; moving the mouse shouldn't hide or affect the key focus, and vice versa. That's been lost with this patch.
- We should hide the emdash unless we have a title:
   
- The truncation seems a little bit on the conservative side, particularly for the URL part. Is it possible to fill the width of the popup more, or does that lead to overflow issues for items with long title and long url? Are we actually aiming to de-emphasize the parts of the URL beyond the domain name?
   
- Seems like the bookmark star icon is always shown with the lil mac result. For consistency, probably best to only show the star next to bookmarked things, or show a different kind of icon next to results that come from our servers (might be a UX question).
   
- It doesn't seem like switch-to-tab or bookmarked results are getting styled as such (notice that foo.com is open in the other tab in this screenshot):
  
  - _Edited_: it looks like the problem is the funky data being returned by the addon: two results are returned for the foo.com domain, but the bookmark action isn't returned both times. I've filed https://github.com/mozilla/universal-search-addon/issues/143 to address this.

```
{"url":"moz-action:switchtab,http://www.foo.com/","image":null,"title":"Foo.com","type":"action bookmark","text":"foo"},
{"url":"http://foo.com/","image":null,"title":"Foo.com","type":"favicon","text":"foo"}
```

**Comment by chuckharmston** [comment]
Thanks for the feedback! I'm definitely doing some refactoring in ways that makes sense to me, but if any of it doesn't jive with you, feel free to tell me to change anything back and I'd be happy to. And, as I mentioned, I'm doing a pretty poor job of organizing my work as I poke around in here, so there are probably 10 commits worth of work in this PR. That definitely makes some of the changes a little opaque. Let me know if there's anything that doesn't make sense.

> The image in the tiny machine results is obscured by the key or click focus:

Ah, I did a `z-index: -1` on the image so its parent's inset `box-shadow` overlaid it. That places it under the background color in the stacking order. I'll try a few other approaches and fix that for this PR, it's pretty busted. Good catch.

> Broken images are, sadly, prominent in the UI when they happen. Maybe we should load the image with JS, and only show it if it loads ok?

Already filed a bug for that: https://github.com/mozilla/universal-search-content/issues/78

> The mouse focus and keyboard focus are intended to be independent; moving the mouse shouldn't hide or affect the key focus, and vice versa. That's been lost with this patch.

I don't think I touched anything that would have changed that; and I don't recall seeing that during my testing. I have a bad merge resolution locally right now, but I'll fix that up and investigate in the morning.

> We should hide the emdash unless we have a title

Agreed, see https://github.com/mozilla/universal-search-addon/issues/142 and https://github.com/mozilla/universal-search-content/issues/80

> The truncation seems a little bit on the conservative side, particularly for the URL part. Is it possible to fill the width of the popup more, or does that lead to overflow issues for items with long title and long url? Are we actually aiming to de-emphasize the parts of the URL beyond the domain name?

I did a percentage max width on both the title and URL, to make sure that long titles still leave room for at least the domain name of the URL. This works really well in some cases, but poorly in others (short title, long URL is particularly awkward). Maybe we could add a percentage `min-width` for the URL, make sure `flex-wrap: nowrap` is set, and let flexbox do the calculations for everything? I'll give that a shot in the morning.

> Seems like the bookmark star icon is always shown with the lil mac result. For consistency, probably best to only show the star next to bookmarked things, or show a different kind of icon next to results that come from our servers (might be a UX question).

That actually came from @bryanbell; the hollow star is for recommendations, the filled star is for bookmarks.

> It doesn't seem like switch-to-tab or bookmarked results are getting styled as such (notice that foo.com is open in the other tab in this screenshot):

Hmmm, that's odd. I'm able to reliably get them styled appropriately; see screenshots above. Do you have reproducible STR?

**Comment by chuckharmston** [comment]
Ah, I bet that last bit is this bug: https://github.com/mozilla/universal-search-addon/issues/143 The way I started slicing results for blending may have surfaced it. 

A few related bugs that, if addressed, should more or less fix the problem: https://github.com/mozilla/universal-search-content/issues/61, https://github.com/mozilla/universal-search-content/issues/72

What's not clear to me is where these fixes should happen. I imagine that we'd want to keep the content iframe as thin as possible, and move as much logic to the server/addon as possible?

**Comment by pdehaan** [comment]
Unrelated, but I can't help but notice a large number of these screenshots above have unparsed `<b>` tags in them. Not sure if we need to tweak our templating escaping to allow for HTML markup, or else find some library to strip out the HTML markup.

**Comment by chuckharmston** [comment]
Definitely. There's a bug filed against tiny machine to strip that out.

**Comment by jaredhirsch** [comment]
Oh, here's a comment I typed out earlier today, but never hit the 'comment' button:

> What's not clear to me is where these fixes should happen. I imagine that we'd want to keep the content iframe as thin as possible, and move as much logic to the server/addon as possible?

Actually, we want to keep as much of the code as possible on the iframe side. Reasons: it's just easier and faster to get stuff done in 'normal' web tech; updating the iframe is quicker than the addon; and (most exciting), if the code's in a webpage, then we could easily run it on a server, for a connected, multi-device, cloud-based future version of search :-)

**Comment by chuckharmston** [comment]
> Actually, we want to keep as much of the code as possible on the iframe side. Reasons: it's just easier and faster to get stuff done in 'normal' web tech; updating the iframe is quicker than the addon; and (most exciting), if the code's in a webpage, then we could easily run it on a server, for a connected, multi-device, cloud-based future version of search :-)

Generally I agree, but is that true even when we're talking about filtering queries out of Places?

**Comment by chuckharmston** [comment]
Alright Mr. @6a68, I think I've cleaned up everything we discussed yesterday. Mind a quick second pass to make sure everything is to your liking?

**Comment by chuckharmston** [comment]
One more fix: I previously forced a square aspect ratio for images in Little Machine results. No longer. The result is less than ideal, but I think it's as good as we're going to get for the content we have for some sites.

## Before

screen shot 2015-11-25 at 8 31 00 am

## After

screen shot 2015-11-25 at 8 30 47 am

**Comment by chuckharmston** [comment]
I'm moving pretty quickly, and this is such a behemoth PR that it's getting in my way. I'm going to put my Python hat on, merge this, it and ask for forgiveness if I break something.

**Pull request merged**